third_party/machinarium/sources/channel.c

```diff
@@ -48,13 +48,39 @@ void mm_channel_write(mm_channel_t *channel, mm_msg_t *msg)
 		mm_sleeplock_unlock(&channel->lock);
 		if (event_mgr_fd > 0)
 			mm_eventmgr_wakeup(event_mgr_fd);
-		return;
+		return MM_OK_RETCODE;
+	}
+
+	switch (channel->limit_policy) {
+	case MM_CHANNEL_UNLIMITED:
+		break;
+	case MM_CHANNEL_LIMIT_HARD: {
+		if (channel->msg_list_count >= channel->chan_limit) {
+			machine_msg_free((machine_msg_t *)msg);
+			mm_sleeplock_unlock(&channel->lock);
+			return MM_NOTOK_RETCODE;
+		}
+	} break;
+	case MM_CHANNEL_LIMIT_SOFT: {
+		// probability of not assepting message is 0 when channel->msg_list_count < channel->chan_limit
+		// probability of not assepting message is 1 when channel->msg_list_count >= 2 * channel->chan_limit
+		// else uniform distribution probability
+		if (machine_lrand48() % channel->chan_limit >
```

Comment:
Can we have a fast path for channel->msg_list_count < channel->chan_limit ?
machine_lrand48() call can be costly